File: LeetCode/79-word-search.py
###############################################################################################
# dfs典型
###########
# 时间复杂度：O((4^le)*mn), le is length of word
# 空间复杂度：O(m*n), st consumes
###############################################################################################
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        le = len(word)
        st = set()
        def dfs(x, y, i): # i represents already matched
            st.add((x,y)) # represents already walked
            if i == le-1:
                return True
            ret, dirs = 0, [[-1, 0], [0, 1], [1, 0], [0, -1]]
            for ux, uy in dirs:
                nx, ny = x+ux, y+uy
                if (nx, ny) not in st and 0 <= nx < m and 0 <= ny < n and board[nx][ny] == word[i+1]:
                    ret |= dfs(nx, ny, i+1)
                    st.remove((nx,ny))
            return ret
        for i in range(m):
            for j in range(n):
                st = set()
                if board[i][j] == word[0] and dfs(i, j, 0):
                    return True
        return False

File: LeetCode/79-word-search.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// easy
////////////////
// Time Complexity: O(m*n*(4**word.length))
// Space Complexity: O(word.length)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    bool dfs(vector<vector<char>>& board, int i, int j, string& word) {
        if (!word.size()) return true;
        // out of boundary or not matched
        if (i<0 || i>=board.size() || j<0 || j>=board[0].size() || board[i][j] != word[0]) return false;
        
        char c = board[i][j];
        // set to walked
        board[i][j] = '*';
        // left string need to match
        string s = word.substr(1);
        // go up, down, left, right
        bool ret = dfs(board, i-1, j, s) || dfs(board, i+1, j, s) || dfs(board, i, j-1, s) || dfs(board, i, j+1, s);
        // set back to not walked
        board[i][j] = c;
        return ret;
    }
    bool exist(vector<vector<char>>& board, string word) {
        for (int i = 0; i < board.size(); i++) 
            for (int j = 0; j < board[0].size(); j++) 
                if (dfs(board, i, j, word)) return true;
        return false;
    }
};